src/game_events/handlers.cpp

```diff
@@ -241,10 +241,24 @@ struct filter_attack : public event_filter {
 	{
 		const unit_map& units = resources::gameboard->units();
 		const auto& loc = first_ ? event_info.loc1 : event_info.loc2;
-		auto unit = units.find(loc);
-		if(unit != units.end() && loc.matches_unit(unit)) {
-			auto attack = event_info.data.optional_child(first_ ? "first" : "second");
-			return swf_.empty() || matches_special_filter(attack, swf_);
+		const auto& loc_d = first_ ? event_info.loc2 : event_info.loc1;
+		auto unit_a = units.find(loc);
+		auto unit_d = units.find(loc_d);
+		if(unit_a != units.end() && loc.matches_unit(unit_a)) {
+			const unit& u = *unit_a;
+			auto temp_weapon = event_info.data.optional_child(first_ ? "first" : "second");
+			const_attack_ptr attack = std::make_shared<const attack_type>(*temp_weapon);
+			if(unit_d != units.end() && loc_d.matches_unit(unit_d)) {
+				const unit& opp = *unit_d;
+				auto temp_other_weapon = event_info.data.optional_child(!first_ ? "first" : "second");
+				const_attack_ptr second_attack = std::make_shared<const attack_type>(*temp_other_weapon);
+				auto ctx = attack->specials_context(u.shared_from_this(), opp.shared_from_this(), loc, loc_d, first_, second_attack);
+				auto opp_ctx = second_attack->specials_context(opp.shared_from_this(), u.shared_from_this(), loc_d, loc, !first_, attack);
```

Comment:
By putting the `shared_from_this()` on the lines declaring `u` and `opp`
```
				const auto opp = unit_d->shared_from_this();
```
then this line becomes the more readable
```
				auto opp_ctx = second_attack->specials_context(opp, u, loc_d, loc, !first_, attack);
```

This one's an optional change - I'd be happy to merge without it.

Reply:
doone, but wait if check complete without fail